### DISCUSSION TITLE
Is there a matcher that simultaneously asserts the type of an object and casts it?

### DISCUSSION THREAD

**Discussion by amomchilov** [Quick/Nimble#928] [Q&A]
- [x] I have read [CONTRIBUTING](https://github.com/Quick/Nimble/blob/main/CONTRIBUTING.md) and have done my best to follow them.

I'm trying to make an assertion about the type of an error, *and* of a field on it. This requires me to cast.

This was my work-around:

```swift
expect(try somethingThatThrows()).to(throwError { error in
    expect(error).to(beAnInstanceOf(MyCustomError))
    expect((error as! MyCustomError).cause) == POSIXError(.ENOMEM)
}
```

Notice how I needed to force cast from `Error` to `MyCustomError` in order to gain access to the `cause` property. This force-cast shouldn't ever fail (because of the typecheck on the previous line), but it's still a little crufty.

Is there a build in matcher that can help with this?

Here are some other alternatives I considered:

```swift
expect(try somethingThatThrows()).to(throwError { error in
    if let error = error as? MyCustomError {
         expect(error.cause) == POSIXError(.ENOMEM)
    } else {
        fail()
    }
```

```swift
expect(try somethingThatThrows()).to(throwError { error in
    expect((error as? MyCustomError).cause) == POSIXError(.ENOMEM)
}
```

### Environment

List the software versions you're using:

- Quick: *4.0.0*
- Nimble: *9.2.0*
- Xcode Version: *13.0 (13A5201i)*
- Swift Version: *5.5*

Please also mention which package manager you used and its version. Delete the
other package managers in this list:

- Swift Package Manager *5.5.0*

**Comment by ikesyo** [comment]
- https://github.com/Quick/Nimble/blob/2c44aeee7eaade06cacb86b2a8261055f423efcf/Sources/Nimble/Matchers/ThrowError.swift#L138-L141
- https://github.com/Quick/Nimble/blob/2c44aeee7eaade06cacb86b2a8261055f423efcf/Tests/NimbleTests/Matchers/ThrowErrorTest.swift#L55-L59

I think this overload is what you want.

```swift
expect(try somethingThatThrows()).to(throwError(errorType: MyCustomError.self) { error in
    expect(error.cause) == POSIXError(.ENOMEM)
}
```

**Reply by amomchilov** [reply]
Oh yeah, that's perfect! Thanks.

(I'm trying to click "Mark as answer", but nothing is happening :/)
